### PR TITLE
Generic loss implementation

### LINKED ISSUE #5: Refactor Loss Functions to Implement a Generic Trait
# Refactor Loss Functions to be Generic

## Description

The current implementation of the `MSE` (Mean Squared Error) loss function in the code is specific to that particular loss function. As our deep learning library grows, we want to add support for other loss functions, such as cross-entropy loss. We should refactor the loss functions to implement a generic trait.

## Proposed Solution

1. Create a new trait called `LossFunction` with the following signature:

```rust
pub trait LossFunction {
   type A: DMatrix<f64>, // Model prediction
   type Y: DMatrix<f64>, // Desired output
   type N: usize, // Batch size
   type C: usize, // Number of features in each sample
   type l_N: DMatrix<f64>, // Column vector of ones of size N (N x 1)
   typel_C: DMatrix<f64> // Column vector of ones of size C (C x 1)
   
   fn forward(&self, y_pred: &DMatrix<f64>, y_true: &DMatrix<f64>) -> f64;
   fn backward(&self, y_pred: &DMatrix<f64>, y_true: &DMatrix<f64>) -> DMatrix<f64>;
}
```
2. Refactor the existing MSE loss function to implement the `LossFunction` trait.
3. Update the NeuralNetwork struct to accept a generic loss function type that implements the `LossFunction` trait.
```rust
pub struct NeuralNetwork {
    pub layers: Vec<Box<Linear>>,
    pub activations: Vec<Box<dyn ActivationFunction>>, // Need dyn dispatch because they're traits
    pub loss: Box<dyn LossFunction>,
}
```
## Implementation Details

- Update the documentation and add relevant tests for the new LossFunction trait and the refactored MSE implementation.
- Implement other common loss functions, such as cross-entropy loss, hinge loss, or any custom loss functions required by your use cases, and ensure they implement the LossFunction trait.
- Consider adding convenience functions or constructors to make it easier to create instances of common loss functions.

## References

- [Rust Traits](https://doc.rust-lang.org/book/ch10-02-traits.html)
- [Generics in Rust](https://doc.rust-lang.org/book/ch10-01-syntax.html)

### PR DESCRIPTION
**Pull request by csuraparaju**
Adds feature described in #5.

### CODE REVIEW

**Comment by tarun-bandi** [comment]
Awesome work!
